blog/user_roles.md

```diff
@@ -0,0 +1,294 @@
+# Real World Phoenix
+## User Roles
+
+Welcome back! In the last episode of this series we implemented user authentication.
+Now let's see how we can implement user authorization by implementing different user roles into our system.
+
+We used Pow! for our authentication and that library also has a guide for implementing a user roles system.
+Let's see how far that get's us and I'd like to actually take it a bit further and implement different
+sign up flows for these different user roles.
+
+But as always, we shouldn't get ahead of ourselves and make it too complicated too soon, so we'll first
+implement basic user roles. We are going to create these users in a simple manner by just adding a selection
+before the registration form that will set this field for the user. I already know I want to be able to define
+multple roles in the future so I'll take that into account when creating the field I'm going to use, but we'll
+not complicate things too much so 1 role can be selected when signing up.
+
+### Migration and schema
+
+We'll need to store the user roles, so we'll add a migration to store a list of users in the users table:
+
+```elixir
+defmodule StudentManager.Repo.Migrations.AddRolesToUsers do
+  use Ecto.Migration
+
+  def change do
+    alter table(:users) do
+      add :roles, {:array, :string}, default: ["student"]
+    end
+  end
+end
+```
+
+In this migration I'm taking advantage of the Postgres Array type to store the list of users in the db. It is
+of course also possible to store the roles in a different table and at those as an association, but for now
+this will do and we can just validate the allowed role values in the Changeset we'll use to store and update
+this user role info.
+
+We'll also need to add this new field to our User struct to be able to use it in our application. Here I have also added the validation of the user role values. 
+
+```elixir
+defmodule StudentManager.Accounts.User do
+  use Ecto.Schema
+  use Pow.Ecto.Schema
+
+  schema "users" do
+    field(:roles, {:array, :string}, default: ["student"])
+
+    pow_user_fields()
+
+    timestamps()
+  end
+
+  def changeset_role(user_or_changeset, attrs) do
+    user_or_changeset
+    |> Changeset.cast(attrs, [:roles])
```

Comment:
I think this is a little dangerous, as it quite easily could lead to users changing their own role if you don't white/black list attributes somewhere higher up.

We've gone a bit back and forth on this on other projects and I still haven't found the perfect solution, but for now I prefer to lock it down already at this level. For example with explicit changeset functions:

```ex
def student_role_changeset(user) do
  change(user, %{roles: ["student"]})
end

def teacher_role_changeset(user) do
  change(user, %{roles: ["teacher"]})
end
```

WDYT?

Reply:
yes, I was actually doing something similar already, but I did leave in one line that exposes this vulnerability indeed. So that shouldn't be there. See below

```elixir
  def changeset(user_or_changeset, attrs) do
    user_or_changeset
    |> pow_changeset(attrs)
    |> changeset_role(attrs) # <= **this is dangerous!**
  end

  def teacher_changeset(user_or_changeset, attrs) do
    user_or_changeset
    |> changeset(attrs)
    |> changeset_role(%{roles: ["teacher"]})
  end

  def student_changeset(user_or_changeset, attrs) do
    user_or_changeset
    |> changeset(attrs)
    |> changeset_role(%{roles: ["student"]})
  end

  def changeset_role(user_or_changeset, attrs) do
    user_or_changeset
    |> cast(attrs, [:roles])
    |> validate_subset(:roles, ~w(student admin teacher parent))
  end
```

However, your solution with the `change/2` function is actually much more succinct. The benefit of the `changeset_role/2`, could be to validate things entered, but in this case that is actually not needed. It could be needed at a later stage if an admin is allowed to update all of the roles of a user at once, but that might be overkill for now.

What do you think?

Reply:
`change/2` is actually meant for internal use, so there's not really a need to perform validations in theory, since you're fully in control of what you pass along with it.

> Wraps the given data in a changeset or adds changes to a changeset.
>
> changes is a map or keyword where the key is an atom representing a field, association or embed and the value is a term. Note the value is directly stored in the changeset with no validation whatsoever. For this reason, this function is meant for working with data internal to the application.

See https://hexdocs.pm/ecto/Ecto.Changeset.html#change/2.

However you can still validate if you want, for example like this:

```ex
def student_role_changeset(user) do
  user
  |> change(%{roles: ["student"]})
  |> validate_roles
end
```